src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Interop/JavaScriptExports.cs

```diff
@@ -210,39 +218,28 @@ public static void CompleteTask(JSMarshalerArgument* arguments_buffer)
             ref JSMarshalerArgument arg_1 = ref arguments_buffer[2];// initialized and set by caller
             // arg_2 set by caller when this is SetException call
             // arg_3 set by caller when this is SetResult call
+
             try
             {
-                var holderGCHandle = arg_1.slot.GCHandle;
-                if (IsGCVHandle(holderGCHandle))
+                // when we arrive here, we could assume that all proxies are owned by calling thread
+                var ctx = JSProxyContext.CapturedInstance = JSProxyContext.AssertCurrentContext();
+                var holder = ctx.ReleasePromiseHolder(arg_1.slot.GCHandle);
```

Comment:
what happens if this handle actually belongs to a different thread?

Reply:
dependeds, if this is mono GCHandle it would not find it in `ThreadJsOwnedObjects` and happily continue. 
I could make it throw.